src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs

```diff
@@ -1477,6 +1478,26 @@ public async Task WaitForExitAsync(CancellationToken cancellationToken = default
             finally
             {
                 Exited -= handler;
+
+                await WaitUntilOutputEOF().ConfigureAwait(false);
+            }
+
+            // Wait until output streams have been drained
+            async ValueTask WaitUntilOutputEOF()
+            {
+                // if we have a hard timeout, we cannot wait for the streams
+                if (!cancellationToken.IsCancellationRequested)
+                {
+                    if (_output != null)
+                    {
+                        await _output.WaitUntilEOFAsync().ConfigureAwait(false);
```

Comment:
To Consider: should we pass the cancellation token to the readers? Note that the sync equivalent doesn't do this.

Reply:
Probably.